### PR TITLE
[FIX] mass_mailing: the blank.gif image in the emails always raise a 500 error

### PR DESCRIPTION
**Pull request by std-odoo**
Bug
===
Since 6185f14807bf527652cd928bf040fd7c596f77ca we check the <mail.mail> existence before marking the <mailing.trace> as opened, but since 57ae1b9b8b61f5f4719a8a81e9d0d21fab58cfda we remove the <mail.mail> automatically when we send them.

The result is that this endpoint always raise a 500 error.

To be: the <mail.mail> existence shouldn't be checked in this endpoint (the token is valid for the raw integer id).

Task-3234519

### CODE REVIEW

File: addons/mass_mailing/controllers/main.py [hunk 1]
```diff
@@ -118,8 +118,7 @@ def mailing_unsubscribe_placeholder_link(self, **post):
     @http.route('/mail/track/<int:mail_id>/<string:token>/blank.gif', type='http', auth='public')
     def track_mail_open(self, mail_id, token, **post):
         """ Email tracking. """
-        mail = request.env['mail.mail'].sudo().browse(mail_id).exists()
-        if not mail or not consteq(token, mail._generate_mail_recipient_token()):
+        if not consteq(token, tools.hmac(request.env(su=True), 'mass_mailing-mail_mail-open', mail_id)):
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/controllers/main.py:121]: Issue is that if one modifies the method, this one will break easily ... ok there are some tests but that's not consistent. Purpose of the cleaning was to move the generation into a single method.

Could we make '_generate_mail_recipient_token' accept an optional ID (in stable) and make it model + using a mail_id in master ?

File: addons/mass_mailing/tests/test_mailing_controllers.py [hunk 2]
```diff
@@ -73,22 +73,24 @@ def test_tracking_url_token(self):
...
 
         mail = self._find_mail_mail_wrecord(self.test_contact)
+        mail_id_int = mail.id
+        mail_tracking_url = mail._get_tracking_url()
         mailing_trace = mail.mailing_trace_ids
         self.assertEqual(mail.state, 'sent')
         self.assertEqual(len(mailing_trace), 1)
         self.assertFalse(mailing_trace.open_datetime)
         self.assertEqual(mailing_trace.trace_status, 'sent')
+        mail.unlink()  # the mail might be removed during the email sending
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/tests/test_mailing_controllers.py:83]: Shoudln't we flush to be sure it is removed ? Not sure it impacts the test, thinking of it.

**Comment by std-odoo** [comment]
@tde-banana-odoo Hi :) 

Indeed you are right, I should still use the method :) 

I did an other PR in master to make the argument required

Cheers

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: line about hmac is actually untouched, we just improve the method that "someone" broke to be a model-like method, when the mail has been removed but we still have its ID somewhere.

**Comment by tde-banana-odoo** [comment]
@fw-bot upto saas-16.2

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@fw-bot up to saas-16.2

**Pull request closed without merging**